src/client/auth/x509.rs

```diff
@@ -18,13 +16,31 @@ pub(super) async fn authenticate_stream(
         auth_command_doc.insert("username", username);
     }
 
-    let auth_command = Command::new("authenticate".into(), "$external".into(), auth_command_doc);
+    Command::new("authenticate".into(), "$external".into(), auth_command_doc)
+}
+
+/// Sends the first client message in the X.509 handshake.
+pub(crate) async fn send_client_first(
+    conn: &mut Connection,
+    credential: &Credential,
+) -> Result<CommandResponse> {
+    let command = build_client_first(credential);
 
-    let server_response = conn.send_command(auth_command, None).await?;
+    conn.send_command(command, None).await
+}
+
+/// Performs X.509 authentication for a given stream.
+pub(super) async fn authenticate_stream(
+    conn: &mut Connection,
+    credential: &Credential,
+    server_first: impl Into<Option<CommandResponse>>,
+) -> Result<()> {
+    let server_response = match server_first.into() {
+        Some(server_first) => server_first,
+        None => send_client_first(conn, credential).await?,
+    };
 
-    if !server_response.is_success()
-        || server_response.raw_response.get_str("dbname") != Ok("$external")
-    {
+    if server_response.raw_response.get_str("dbname") != Ok("$external") {
```

Comment:
Due to the lack of `"ok": 1` in the server's speculative response, the cleanest solution is to just rely on the "dbname" field alone to verify that authentication succeeded, as it won't be present in an `"ok": 0` response